sklearn/linear_model/_glm/tests/test_glm.py

```diff
@@ -29,6 +64,480 @@ def regression_data():
+
+    # Same with sample_weight.
+    model = (
+        clone(model).set_params(**params).fit(X, y, sample_weight=np.ones(X.shape[0]))
+    )
+    assert model.intercept_ == pytest.approx(intercept, rel=rtol)
+    assert_allclose(model.coef_, coef, rtol=rtol)
+
+
+@pytest.mark.parametrize("solver", SOLVERS)
+@pytest.mark.parametrize("fit_intercept", [True, False])
+def test_glm_regression_hstacked_X(solver, fit_intercept, glm_dataset):
+    """Test that GLM converges for all solvers to correct solution on hstacked data.
+
+    We work with a simple constructed data set with known solution.
+    Fit on [X] with alpha is the same as fit on [X, X]/2 with alpha/2.
+    For long X, [X, X] is still a long but singular matrix.
+    """
+    model, X, y, _, coef_with_intercept, coef_without_intercept = glm_dataset
+    n_samples, n_features = X.shape
+    alpha = 1.0  # because glm_dataset uses this.
+    params = dict(
+        alpha=alpha / 2,
+        fit_intercept=fit_intercept,
+        # solver=solver,  # only lbfgs available
+        tol=1e-12,
+        max_iter=1000,
+    )
+
+    if solver == "lbfgs" and not fit_intercept:
+        # Sometimes (depending on global_random_seed) lbfgs fails with:
+        # Line search cannot locate an adequate point after MAXLS function and gradient
+        # evaluations. Previous x, f and g restored.
+        # Possible causes: 1 error in function or gradient evaluation;
+        #                  2 rounding error dominate computation.
+        pytest.xfail()
+
+    model = clone(model).set_params(**params)
+    X = X[:, :-1]  # remove intercept
+    X = 0.5 * np.concatenate((X, X), axis=1)
+    assert np.linalg.matrix_rank(X) <= min(n_samples, n_features - 1)
+    if fit_intercept:
+        coef = coef_with_intercept
+        intercept = coef[-1]
+        coef = coef[:-1]
+    else:
+        coef = coef_without_intercept
+        intercept = 0
+    model.fit(X, y)
+
+    rtol = 1e-4
+    assert model.intercept_ == pytest.approx(intercept, rel=rtol)
+    assert_allclose(model.coef_, np.r_[coef, coef], rtol=rtol)
+
+
+@pytest.mark.parametrize("solver", SOLVERS)
+@pytest.mark.parametrize("fit_intercept", [True, False])
+def test_glm_regression_vstacked_X(solver, fit_intercept, glm_dataset):
+    """Test that GLM converges for all solvers to correct solution on vstacked data.
+
+    We work with a simple constructed data set with known solution.
+    Fit on [X] with alpha is the same as fit on [X], [y]
+                                                [X], [y] with 1 * alpha.
+    It is the same alpha as the average loss stays the same.
+    For wide X, [X', X'] is a singular matrix.
+    """
+    model, X, y, _, coef_with_intercept, coef_without_intercept = glm_dataset
+    n_samples, n_features = X.shape
+    alpha = 1.0  # because glm_dataset uses this.
+    params = dict(
+        alpha=alpha,
+        fit_intercept=fit_intercept,
+        # solver=solver,  # only lbfgs available
```

Comment:
It well become handy once there is more than one solver. The parametrization anticipates this.
Are you ok with it or shall I remove it for the time being?

Reply:
As always, it's a bit tricky. For simplicity, I'll simply remove the solver parameter/parameterisation.

`_GeneralizedLinearRegressor` has a `solver` parameter, but the public estimators like `PoissonRegressor` don't. I still prefer to test the public estimators instead of the private one as it is closer to an end-to-end test.

Reply:
I see, is this proposal sensible regarding what you mentioned ?

```suggestion
        # As `solver` is available for a few estimators and
        # as "lbfgs" is the only solver, we left it commented.
        # TODO: make sure `solver` is exposed for all the
        # estimators class subclassing `_GeneralizedLinearRegressor`
        # and uncomment this line.
        # solver=solver,
```

Reply:
In 28fdbab80f7582510f77a1fb397a28558be8b050 I added
```
        # While _GeneralizedLinearRegressor exposes the solver parameter, public
        # estimators currently do not, and lbfgs is the only solver anyway.
        # TODO: Expose solver as soon as we have a second solver to choose from.
        # solver=solver,  # only lbfgs available
```